keps/sig-node/0028-node-readiness++.md

```diff
@@ -0,0 +1,237 @@
+    Quantity: x // Sample fields for illustration purposes
+    UnHealthy: y 
+    DeviceIDs []string
+  }
+}
+
+```
+
+One of the goals for this proposal is to evaluate the viability and usefullness of combining Status and Health for extensions.
+
+### Approach B - Use Readiness of Node extension pods
+
+This approach assumes that all node level extensions will get deployed as kubernetes pods and that the health of those extension pods can be exposed via existing Readiness probes.
+
+Similar to the previous approach `A`, a special taint will be used to track the overall readiness of the node and the node controller and kubelet will manage the special taint.
+
+The Node Object will be extended to include a list of system extensions that are identified via a label selector and a namespace.
+
+```yaml
+type Node struct {
+ …
+ Spec:
+  SystemExtensions []SystemExtension
+...
+}
+
+type SystemExtension struct {
+   Selector LabelSelector
+   Namespace Namespace
+   AffectsReadiness bool // Ignore non-critical node extensions from node readiness
+}
+```
+
+Node selector will take into account `Readiness` all the pods that match the selector(s) specified in the Node Spec `SystemExtensions` field to determine Node Readiness.
+
+The individual system extensions are required to detect kubelet crash (or restart) quickly and reflect that in their Readiness.
+If the Node Controller observes that all the system extension pods are Ready after the kubelet places the special meta readiness taint, it will remove that taint immediately.
+If any of the pods become unready, then the node controller will place the meta readiness taint on the nodes.
+
+This approach has the advantage of not introducing additional API concepts to track state as illustrated in Approach A.
+Side-car containers can be used to expose Readiness more easily than in approach A where it involves updating API objects directly.
+
+Handling of kubelet restarts can be tricky though since there is a chance that the extension pods may not reflect their connections with the kubelet soon after the kubelet restarts.
+
+### Approach C - Use conditions
+
+This design continues the existing pattern of expressing health of various node features and components via Conditions.
+For example the Kubelet uses conditions heavily and the Node Problem Detector also uses conditions. Conditions allow for recording heartbeats already.
+
+This design is similar to the approach `A` where instead of recording state to a separate object, kubelet and other system components will record their state via Conditions.
+The node controller will be extended to support a configurable policy on what Conditions influence the meta readiness of a node.
+
+The kubelet and node controller will use a taint similar to the one illustrated in approach `A` to signal node meta readiness.
+
+The scheduler needs to understand that certain special pods (system pods) can be scheduled to nodes even if the node isn’t logically ready.
```

Comment:
This is true for all solutions, isn't it? Essentially, this feature:

* Defines two separate readiness tiers:
    - Ready - existing status indicator, this is more "kubelet ready" or "node core ready" or "ready for control plane workloads" or "ready for system workloads"
    - MetaReadiness - the new state, which represents "ready for user workloads".
 * Provides a mechanism to to stall user workloads (the taint) while system workloads are still coming up. system workloads need to tolerate those taints.

It might be helpful to consider each of these as separate problems that need solving:
1. How do we differentiate "system workload" and "user workload" readiness in a way that works for scheduling? All solutions currently discussed use existing readiness as "system workload readiness" and a taint as "not ready for user workloads", from what I can see.
2. How do we flip each of those readiness flags? This could vary a little based on the solution to (1) - for example the current proposal does not require any changes to handle the "system workload readiness" flag, IIUC. For the "user workload readiness", this can be a stored component-level status + a policy (hardcoded policy in this case using AffectsReadiness) - e.g., "A". Alternatively it could use pod status ("B"), or conditions ("C").

Re-organizing the doc in terms of these may make it easier digest.

Reply:
> This is true for all solutions, isn't it?

All the other approaches uses a taint to express "MetaReadiness" which means no changes are needed to the scheduler. Conditions on the other hand need special logic to be embedded into the scheduler and other controllers like autoscaler.